src/coreComponents/physicsSolvers/multiphysics/LagrangianContactSolver.cpp

```diff
@@ -1314,65 +1314,111 @@ void LagrangianContactSolver::
       rotationMatrix = subRegion.getReference< array3d< real64 > >( viewKeyStruct::rotationMatrixString );
       arrayView2d< localIndex const > const & elemsToFaces = subRegion.faceList();
 
-      forAll< parallelHostPolicy >( subRegion.size(), [=] ( localIndex const kfe )
-      {
-        localIndex const numNodesPerFace = faceToNodeMap.sizeOfArray( elemsToFaces[kfe][0] );
 
-        globalIndex rowDOF[12];
-        real64 nodeRHS[12];
-        stackArray2d< real64, 3*4*3 > dRdT( 3*numNodesPerFace, 3 );
-        globalIndex colDOF[3];
-        for( localIndex i = 0; i < 3; ++i )
-        {
-          colDOF[i] = tracDofNumber[kfe] + i;
-        }
+//      finiteElement::FiniteElementBase const &
+//      fe = subRegion.getReference< finiteElement::FiniteElementBase >( this->getDiscretizationName() );
+//      finiteElement::dispatch2D( fe,
+//                                 [&] ( auto const finiteElement )
+//      {
+//        using FE_TYPE = TYPEOFREF( finiteElement );
+//
+//        constexpr localIndex numNodesPerElem = FE_TYPE::numNodes;
+//        constexpr localIndex numQuadraturePointsPerElem = FE_TYPE::numQuadraturePoints;
```

Comment:
@castelletto1 @af1990 
I meant to note this, but to use a dispatch, we would need to know the element type for all the elements in the kernel. Since we are using faces, and the faces are NOT organized by shape, we simply can't do this...resulting in the unholy mess below. I haven't been able to think of a way out of this limitation that will extend to large strain. For small strain only, the face elements could be organized by shape if we wanted, which would clean this up nicely. However, for the general finite deformation case...well...I think at that point we really can't program with any knowledge shapes of contacting faces shapes (both). The biggest problem I can see is that the discretization method of constraint enforcement would dictate what the algorithmic pattern is...and those patterns can vary widely from method to method (mortar, symmetric mortar, common plane). 

Point being, @castelletto1 I/we have an obvious problem with the mesh infrastructure for integration over surface faces since we don't know the geometry of the face. The presence of branching in the kernel is typically bad, but I don't know how much time we would be spending in this kernel. @af1990 Any caliper estimates of this? The solver time should be dominating, so a percentage relative to the non-solver time percentage would be also informative.

Reply:
I have already tried to use caliper in the past, but I am running on a simple desktop and times are quite unreliable.

Reply:
```
Path                                             Min time/rank Max time/rank Avg time/rank Time %    
main                                                  0.009661      0.009661      0.009661  0.042483 
  geosx::EventManager::Run                            0.001538      0.001538      0.001538  0.006763 
    geosx::SolverBase::Execute                        0.006046      0.006046      0.006046  0.026586 
      geosx::Communication~~s::SynchronizeUnpack      0.000005      0.000005      0.000005  0.000022 
      geosx::Communication~~chronizePackSendRecv      0.000291      0.000291      0.000291  0.001280 
      geosx::Communication~~izePackSendRecvSizes      0.000074      0.000074      0.000074  0.000325 
      geosx::LagrangianCon~~onlinearImplicitStep      0.536631      0.536631      0.536631  2.359761 
        geosx::LagrangianCo~~StateForElasticStep      0.000008      0.000008      0.000008  0.000035 
        geosx::SolidMechani~~teToBeginningOfStep      0.000036      0.000036      0.000036  0.000158 
          geosx::SolidMechan~~sToBeginningOfStep      0.000171      0.000171      0.000171  0.000752 
        geosx::LagrangianCo~~UpdateFractureState      0.001412      0.001412      0.001412  0.006209 
          geosx::Communicati~~:SynchronizeUnpack      0.000021      0.000021      0.000021  0.000092 
          geosx::Communicati~~ronizePackSendRecv      0.000701      0.000701      0.000701  0.003083 
          geosx::Communicati~~ePackSendRecvSizes      0.000222      0.000222      0.000222  0.000976 
        geosx::LagrangianCo~~ApplySystemSolution      0.089336      0.089336      0.089336  0.392843 
          geosx::LagrangianC~~mputeFaceNodalArea      0.050347      0.050347      0.050347  0.221394 
          geosx::Communicati~~:SynchronizeUnpack      0.000140      0.000140      0.000140  0.000616 
          geosx::Communicati~~ronizePackSendRecv      0.002476      0.002476      0.002476  0.010888 
          geosx::Communicati~~ePackSendRecvSizes      0.000991      0.000991      0.000991  0.004358 
          geosx::SolidMechan~~pplySystemSolution      0.020655      0.020655      0.020655  0.090828 
            geosx::Communicat~~SynchronizeUnpack      0.000148      0.000148      0.000148  0.000651 
            geosx::Communicat~~onizePackSendRecv      0.003431      0.003431      0.003431  0.015087 
            geosx::Communicat~~PackSendRecvSizes      0.001915      0.001915      0.001915  0.008421 
        geosx::LagrangianCo~~Solver::SolveSystem      0.000981      0.000981      0.000981  0.004314 
          geosx::SolverBase::SolveSystem             17.710207     17.710207     17.710207 77.878215 
        geosx::LagrangianCo~~lculateResidualNorm      0.004919      0.004919      0.004919  0.021631 
        geosx::LagrangianCo~~yBoundaryConditions      0.000448      0.000448      0.000448  0.001970 
          geosx::SolidMechan~~BoundaryConditions      0.001242      0.001242      0.001242  0.005462 
            geosx::SolidMecha~~cementBC_implicit      0.000492      0.000492      0.000492  0.002164 
              geosx::FieldSpec~~onManager::Apply      0.052812      0.052812      0.052812  0.232234 
            geosx::SolidMecha~~RSApplyTractionBC      0.000897      0.000897      0.000897  0.003944 
              geosx::FieldSpec~~onManager::Apply      0.010453      0.010453      0.010453  0.045966 
            geosx::FieldSpeci~~ionManager::Apply      0.008792      0.008792      0.008792  0.038662 
        geosx::LagrangianCo~~ver::AssembleSystem      0.011275      0.011275      0.011275  0.049580 
          geosx::LagrangianC~~embleStabilization      1.311280      1.311280      1.311280  5.766175 
            geosx::Lagrangian~~puteFaceNodalArea      0.170059      0.170059      0.170059  0.747811 
          geosx::LagrangianC~~acementAndTraction      0.257537      0.257537      0.257537  1.132484 
            geosx::Lagrangian~~puteFaceNodalArea      0.132491      0.132491      0.132491  0.582611 
          geosx::LagrangianC~~ivativeWrtTraction      0.078620      0.078620      0.078620  0.345721 
          geosx::SolidMechan~~EM::AssembleSystem      0.021829      0.021829      0.021829  0.095990 
            geosx::SolidMecha~~M::AssemblyLaunch      0.001944      0.001944      0.001944  0.008548 
              geosx::SolidMech~~ontactConstraint      0.000299      0.000299      0.000299  0.001315 
              geosx::finiteEle~~ernelApplication      0.001647      0.001647      0.001647  0.007242 
                3>::kernelLaunch                      1.530952      1.530952      1.530952  6.732152 
              geosx::SolidMech~~oBeginningOfStep      0.006858      0.006858      0.006858  0.030157 
          geosx::Communicati~~:SynchronizeUnpack      0.000176      0.000176      0.000176  0.000774 
          geosx::Communicati~~ronizePackSendRecv      0.005469      0.005469      0.005469  0.024049 
          geosx::Communicati~~ePackSendRecvSizes      0.001965      0.001965      0.001965  0.008641 
      geosx::SolverBase::SetupSystem                  0.038958      0.038958      0.038958  0.171312 
        geosx::Communicatio~~::SynchronizeUnpack      0.000003      0.000003      0.000003  0.000013 
        geosx::Communicatio~~hronizePackSendRecv      0.000115      0.000115      0.000115  0.000506 
        geosx::Communicatio~~zePackSendRecvSizes      0.000078      0.000078      0.000078  0.000343 
        geosx::LagrangianCo~~ctSolver::SetupDofs      0.000659      0.000659      0.000659  0.002898 
          geosx::Communicati~~:SynchronizeUnpack      0.000002      0.000002      0.000002  0.000009 
          geosx::Communicati~~ronizePackSendRecv      0.000113      0.000113      0.000113  0.000497 
          geosx::Communicati~~ePackSendRecvSizes      0.000048      0.000048      0.000048  0.000211 
          geosx::SolidMechan~~gianFEM::SetupDofs      0.000967      0.000967      0.000967  0.004252 
            geosx::Communicat~~SynchronizeUnpack      0.000010      0.000010      0.000010  0.000044 
            geosx::Communicat~~onizePackSendRecv      0.000125      0.000125      0.000125  0.000550 
            geosx::Communicat~~PackSendRecvSizes      0.000060      0.000060      0.000060  0.000264 
      geosx::LagrangianCon~~ComputeFaceNodalArea      0.002454      0.002454      0.002454  0.010791 
      geosx::LagrangianCon~~r::ComputeTolerances      0.003865      0.003865      0.003865  0.016996 
      geosx::LagrangianCon~~puteRotationMatrices      0.000297      0.000297      0.000297  0.001306 
    geosx::SurfaceGenerator::SeparationDriver         0.036529      0.036529      0.036529  0.160631 
      geosx::ObjectManagerBase::FixUpDownMaps         0.000034      0.000034      0.000034  0.000150 
      geosx::Communication~~s::SynchronizeUnpack      0.000001      0.000001      0.000001  0.000004 
      geosx::Communication~~chronizePackSendRecv      0.000012      0.000012      0.000012  0.000053 
      geosx::Communication~~izePackSendRecvSizes      0.000009      0.000009      0.000009  0.000040 
    geosx::SiloOutput::Execute                        0.584046      0.584046      0.584046  2.568262 
  geosx::ProblemManager::ProblemSetup                 0.003111      0.003111      0.003111  0.013680 
    geosx::FluxApproximat~~nditions_PreSubGroups      0.000234      0.000234      0.000234  0.001029 
      geosx::FieldSpecificationManager::Apply         0.000010      0.000010      0.000010  0.000044 
    geosx::LagrangianCont~~itializeFractureState      0.000007      0.000007      0.000007  0.000031 
    geosx::FieldSpecificationManager::Apply           0.000092      0.000092      0.000092  0.000405 
    geosx::FiniteElementD~~hapeFunctionGradients      0.000234      0.000234      0.000234  0.001029 
    geosx::ProblemManager::GenerateMesh               0.000428      0.000428      0.000428  0.001882 
      geosx::DomainPartiti~~:SetupCommunications      0.000092      0.000092      0.000092  0.000405 
        geosx::FaceManager::SortAllFaceNodes          0.000278      0.000278      0.000278  0.001222 
        geosx::CommunicationTools::FindGhosts         0.000060      0.000060      0.000060  0.000264 
          geosx::ObjectManag~~ase::FixUpDownMaps      0.000004      0.000004      0.000004  0.000018 
        geosx::Communicatio~~tionBoundaryObjects      0.000046      0.000046      0.000046  0.000202 
        geosx::Communicatio~~AssignGlobalIndices      0.000246      0.000246      0.000246  0.001082 
          geosx::EdgeManager~~GlobalIndexNumbers      0.000034      0.000034      0.000034  0.000150 
          geosx::ObjectManag~~ctGlobalToLocalMap      0.000101      0.000101      0.000101  0.000444 
          geosx::FaceManager~~GlobalIndexNumbers      0.000032      0.000032      0.000032  0.000141 
      geosx::CellElementRe~~::GenerateAggregates      0.000001      0.000001      0.000001  0.000004 
      geosx::DomainPartition::GenerateSets            0.000388      0.000388      0.000388  0.001706 
      geosx::NodeManager::SetEdgeMaps                 0.000070      0.000070      0.000070  0.000308 
      geosx::EdgeManager::BuildEdges                  0.000472      0.000472      0.000472  0.002076 
        geosx::populateMaps                           0.000050      0.000050      0.000050  0.000220 
        geosx::resizeEdgeToFaceMap                    0.000039      0.000039      0.000039  0.000171 
        geosx::createEdgesByLowestNode                0.000640      0.000640      0.000640  0.002814 
      geosx::NodeManager::SetFaceMaps                 0.000486      0.000486      0.000486  0.002137 
      geosx::FaceManager::BuildFaces                  0.000674      0.000674      0.000674  0.002964 
        geosx::populateMaps                           0.000261      0.000261      0.000261  0.001148 
        geosx::resizeFaceToNodeMap                    0.000202      0.000202      0.000202  0.000888 
        geosx::createFacesByLowestNode                0.000715      0.000715      0.000715  0.003144 
      geosx::NodeManager::SetElementMaps              0.000109      0.000109      0.000109  0.000479 
      geosx::ObjectManager~~ructGlobalToLocalMap      0.000051      0.000051      0.000051  0.000224 
      geosx::InternalMeshGenerator::GenerateMesh      0.001464      0.001464      0.001464  0.006438
```

The total time spent in `ComputeFaceNodalArea` is 1.6% of the runtime. @rrsettgast, your implementation is inside `AssembleForceResidualDerivativeWrtTraction` (0.35%), but the function does also something else.

Reply:
@rrsettgast Is that acceptable or is it too large?
The example has 726 nodes and 100 fracture elements. It runs for 10 time steps.

Reply:
That sounds like a really small problem to me.

Reply:
It was ... This is bigger, with about 50k nodes and 1600 fracture elements. It runs for 4 time steps.
```
Path                                                                             Min time/rank Max time/rank Avg time/rank Time %    
main                                                                                  0.010058      0.026846      0.016332  0.003966 
  geosx::EventManager::Run                                                            0.002919      0.647852      0.280307  0.068068 
    geosx::SolverBase::Execute                                                        0.016736      0.044070      0.030889  0.007501 
      geosx::CommunicationTools::SynchronizeUnpack                                    0.000156      0.000714      0.000359  0.000087 
        geosx::NeighborCommunicator::UnpackBufferForSync                              0.000048      0.000086      0.000062  0.000015 
      geosx::CommunicationTools::SynchronizePackSendRecv                              0.000107      0.000217      0.000148  0.000036 
        geosx::NeighborCommunicator::PackCommBufferForSync                            0.000140      0.000391      0.000277  0.000067 
      geosx::CommunicationTools::SynchronizePackSendRecvSizes                         0.000086      0.000132      0.000110  0.000027 
        geosx::NeighborCommunicator::PackCommSizeForSync                              0.000149      0.000383      0.000273  0.000066 
      geosx::LagrangianContactSolver::NonlinearImplicitStep                           8.691028     14.393871     11.534755  2.801038 
        geosx::LagrangianContactSolver::SetFractureStateForElasticStep                0.000006      0.000010      0.000007  0.000002 
        geosx::SolidMechanicsLagrangianFEM::ResetStateToBeginningOfStep               0.000154      0.000224      0.000185  0.000045 
          geosx::SolidMechanicsLagrangianFEM::ResetStressToBeginningOfStep            0.001853      0.002875      0.002360  0.000573 
        geosx::LagrangianContactSolver::UpdateFractureState                           0.002050      0.002643      0.002488  0.000604 
          geosx::CommunicationTools::SynchronizeUnpack                                0.003550      0.005532      0.004760  0.001156 
            geosx::NeighborCommunicator::UnpackBufferForSync                          0.000485      0.000886      0.000575  0.000140 
          geosx::CommunicationTools::SynchronizePackSendRecv                          0.503532      2.157057      1.217005  0.295531 
            geosx::NeighborCommunicator::PackCommBufferForSync                        0.001041      0.003237      0.002021  0.000491 
          geosx::CommunicationTools::SynchronizePackSendRecvSizes                     0.000673      0.001587      0.000974  0.000236 
            geosx::NeighborCommunicator::PackCommSizeForSync                          0.001123      0.003218      0.002094  0.000509 
        geosx::LagrangianContactSolver::ApplySystemSolution                           0.286413      0.564633      0.439456  0.106715 
          geosx::LagrangianContactSolver::ComputeFaceNodalArea                        0.175432      0.343082      0.270772  0.065753 
          geosx::CommunicationTools::SynchronizeUnpack                                0.013512      0.035087      0.024659  0.005988 
            geosx::NeighborCommunicator::UnpackBufferForSync                          0.002490      0.004990      0.003259  0.000791 
          geosx::CommunicationTools::SynchronizePackSendRecv                          0.009101      0.023824      0.015342  0.003726 
            geosx::NeighborCommunicator::PackCommBufferForSync                        0.005949      0.019953      0.012819  0.003113 
          geosx::CommunicationTools::SynchronizePackSendRecvSizes                     0.002887      0.004893      0.004080  0.000991 
            geosx::NeighborCommunicator::PackCommSizeForSync                          0.006494      0.018560      0.012269  0.002979 
          geosx::SolidMechanicsLagrangianFEM::ApplySystemSolution                     0.028090      0.043431      0.035663  0.008660 
            geosx::CommunicationTools::SynchronizeUnpack                              0.015914      0.033081      0.026243  0.006373 
              geosx::NeighborCommunicator::UnpackBufferForSync                        0.003319      0.005995      0.004470  0.001085 
            geosx::CommunicationTools::SynchronizePackSendRecv                        0.054884      0.205299      0.134524  0.032667 
              geosx::NeighborCommunicator::PackCommBufferForSync                      0.008781      0.016347      0.012660  0.003074 
            geosx::CommunicationTools::SynchronizePackSendRecvSizes                   0.006956      0.010274      0.008507  0.002066 
              geosx::NeighborCommunicator::PackCommSizeForSync                        0.007670      0.011094      0.009903  0.002405 
        geosx::LagrangianContactSolver::SolveSystem                                   0.001139      0.001605      0.001271  0.000309 
          geosx::SolverBase::SolveSystem                                            358.225998    358.368110    358.290197 87.005260 
        geosx::LagrangianContactSolver::CalculateResidualNorm                         3.090381     17.636741     10.362617  2.516402 
        geosx::LagrangianContactSolver::ApplyBoundaryConditions                       0.000508      0.000808      0.000659  0.000160 
          geosx::SolidMechanicsLagrangianFEM::ApplyBoundaryConditions                 0.001196      0.001768      0.001534  0.000372 
            geosx::SolidMechanicsLagrangianFEM::ApplyDisplacementBC_implicit          0.000469      0.000744      0.000631  0.000153 
              geosx::FieldSpecificationManager::Apply                                 0.031280      0.128976      0.082500  0.020034 
            geosx::SolidMechanicsLagrangianFEM::CRSApplyTractionBC                    0.000911      0.001768      0.001213  0.000295 
              geosx::FieldSpecificationManager::Apply                                 0.011028      0.022130      0.015033  0.003651 
            geosx::FieldSpecificationManager::Apply                                   0.007976      0.013819      0.011189  0.002717 
        geosx::LagrangianContactSolver::AssembleSystem                                0.016507      0.024106      0.018853  0.004578 
          geosx::LagrangianContactSolver::AssembleStabilization                       0.052748      8.579885      3.742729  0.908864 
            geosx::LagrangianContactSolver::ComputeFaceNodalArea                      0.880900      1.164383      1.010735  0.122721 
          geosx::LagrangianContactSolver::As~~rivativeWrtDisplacementAndTraction      0.002163      1.938691      0.737897  0.179187 
            geosx::LagrangianContactSolver::ComputeFaceNodalArea                      0.560237      1.069903      0.793809  0.096382 
          geosx::LagrangianContactSolver::As~~ForceResidualDerivativeWrtTraction      0.066978      0.429982      0.244479  0.059368 
          geosx::SolidMechanicsLagrangianFEM::AssembleSystem                          0.665581      1.121341      0.944452  0.229346 
            geosx::SolidMechanicsLagrangianFEM::AssemblyLaunch                        0.003806      0.007710      0.005205  0.001264 
              geosx::SolidMechanicsLagrangianFEM::ApplyContactConstraint              0.000353      0.000427      0.000397  0.000096 
              geosx::finiteElement::regionBasedKernelApplication                      0.003892      0.006055      0.004649  0.001129 
                3>::kernelLaunch                                                     11.878998     21.284316     17.236675  4.185661 
              geosx::SolidMechanicsLagrangianFEM::ResetStressToBeginningOfStep        0.154920      0.275840      0.219526  0.053309 
          geosx::CommunicationTools::SynchronizeUnpack                                0.028919      0.067657      0.052249  0.012688 
            geosx::NeighborCommunicator::UnpackBufferForSync                          0.003637      0.006159      0.004713  0.001145 
          geosx::CommunicationTools::SynchronizePackSendRecv                          0.242880      0.872612      0.513161  0.124613 
            geosx::NeighborCommunicator::PackCommBufferForSync                        0.011241      0.031934      0.021815  0.005297 
          geosx::CommunicationTools::SynchronizePackSendRecvSizes                     0.009588      0.013046      0.010988  0.002668 
            geosx::NeighborCommunicator::PackCommSizeForSync                          0.014955      0.031653      0.023054  0.005598 
      geosx::SolverBase::SetupSystem                                                  0.107922      0.200094      0.156239  0.037940 
        geosx::CommunicationTools::SynchronizeUnpack                                  0.000318      0.000975      0.000621  0.000151 
          geosx::NeighborCommunicator::UnpackBufferForSync                            0.000067      0.000106      0.000081  0.000020 
        geosx::CommunicationTools::SynchronizePackSendRecv                            0.000130      0.000707      0.000340  0.000083 
          geosx::NeighborCommunicator::PackCommBufferForSync                          0.000246      0.000586      0.000434  0.000105 
        geosx::CommunicationTools::SynchronizePackSendRecvSizes                       0.000092      0.000157      0.000125  0.000030 
          geosx::NeighborCommunicator::PackCommSizeForSync                            0.000239      0.000543      0.000413  0.000100 
        geosx::LagrangianContactSolver::SetupDofs                                     0.000585      0.000667      0.000623  0.000151 
          geosx::CommunicationTools::SynchronizeUnpack                                0.000162      0.000708      0.000378  0.000092 
            geosx::NeighborCommunicator::UnpackBufferForSync                          0.000047      0.000078      0.000058  0.000014 
          geosx::CommunicationTools::SynchronizePackSendRecv                          0.000124      0.000243      0.000154  0.000037 
            geosx::NeighborCommunicator::PackCommBufferForSync                        0.000147      0.000407      0.000284  0.000069 
          geosx::CommunicationTools::SynchronizePackSendRecvSizes                     0.000076      0.000124      0.000101  0.000025 
            geosx::NeighborCommunicator::PackCommSizeForSync                          0.000158      0.000405      0.000290  0.000070 
          geosx::SolidMechanicsLagrangianFEM::SetupDofs                               0.004972      0.017155      0.010250  0.002489 
            geosx::CommunicationTools::SynchronizeUnpack                              0.000169      0.000507      0.000302  0.000073 
              geosx::NeighborCommunicator::UnpackBufferForSync                        0.000041      0.000061      0.000047  0.000011 
            geosx::CommunicationTools::SynchronizePackSendRecv                        0.000164      0.000409      0.000258  0.000063 
              geosx::NeighborCommunicator::PackCommBufferForSync                      0.000133      0.000229      0.000191  0.000046 
            geosx::CommunicationTools::SynchronizePackSendRecvSizes                   0.000107      0.000150      0.000131  0.000032 
              geosx::NeighborCommunicator::PackCommSizeForSync                        0.000135      0.000224      0.000184  0.000045 
      geosx::LagrangianContactSolver::ComputeFaceNodalArea                            0.004414      0.008244      0.006378  0.001549 
      geosx::LagrangianContactSolver::ComputeTolerances                               0.000866      0.002252      0.001616  0.000392 
      geosx::LagrangianContactSolver::ComputeRotationMatrices                         0.000149      0.000177      0.000165  0.000040 
    geosx::SurfaceGenerator::SeparationDriver                                         1.855167      1.913966      1.889090  0.458736 
      geosx::ObjectManagerBase::FixUpDownMaps                                         0.001893      0.004457      0.002744  0.000666 
      "Unpack FixedToManyElementRelation"                                             0.000446      0.001140      0.000775  0.000188 
      "Unpack OrderedVariableToManyElementRelation"                                   0.000927      0.001757      0.001310  0.000318 
      geosx::ObjectManagerBase::UnpackGlobalMaps                                      0.004630      0.016809      0.011406  0.002770 
      geosx::ParallelTopologyChange::UnpackNewAndModifiedObjectsOnOwningRanks         0.004598      0.043420      0.018698  0.004540 
        "Unpack FixedToManyElementRelation"                                           0.000077      0.000703      0.000319  0.000077 
        "Unpack OrderedVariableToManyElementRelation"                                 0.000082      0.000644      0.000275  0.000067 
        geosx::ObjectManagerBase::UnpackGlobalMaps                                    0.003446      0.009194      0.005761  0.001399 
      geosx::CommunicationTools::SynchronizeUnpack                                    0.000036      0.000120      0.000084  0.000020 
        geosx::NeighborCommunicator::UnpackBufferForSync                              0.000017      0.000051      0.000039  0.000010 
      geosx::CommunicationTools::SynchronizePackSendRecv                              0.000023      0.249882      0.117641  0.028567 
        geosx::NeighborCommunicator::PackCommBufferForSync                            0.000063      0.000134      0.000097  0.000023 
      geosx::CommunicationTools::SynchronizePackSendRecvSizes                         0.000026      0.000049      0.000037  0.000009 
        geosx::NeighborCommunicator::PackCommSizeForSync                              0.000054      0.000113      0.000078  0.000019 
    geosx::SiloOutput::Execute                                                        1.094462      2.211632      1.710826  0.415448 
  geosx::ProblemManager::ProblemSetup                                                 0.009132      0.015447      0.012208  0.002965 
    geosx::FluxApproximationBase::InitializePostInitialConditions_PreSubGroups        0.000339      0.000495      0.000423  0.000103 
      geosx::FieldSpecificationManager::Apply                                         0.000010      0.000019      0.000015  0.000004 
    geosx::LagrangianContactSolver::InitializeFractureState                           0.000007      0.000012      0.000010  0.000003 
    geosx::FieldSpecificationManager::Apply                                           0.000100      0.000178      0.000138  0.000034 
    geosx::FiniteElementDiscretization::CalculateShapeFunctionGradients               0.004803      0.008033      0.006464  0.001570 
    geosx::ProblemManager::GenerateMesh                                               0.002099      0.016573      0.009261  0.002249 
      geosx::DomainPartition::SetupCommunications                                     0.003530      0.003755      0.003630  0.000881 
        geosx::FaceManager::SortAllFaceNodes                                          0.002748      0.004876      0.003832  0.000930 
        geosx::CommunicationTools::FindGhosts                                         0.003074      0.089062      0.047492  0.011533 
          geosx::ObjectManagerBase::FixUpDownMaps                                     0.002479      0.004417      0.003446  0.000837 
          geosx::NeighborCommunicator::UnpackAndRebuildSyncLists                      0.000461      0.000709      0.000589  0.000143 
          geosx::NeighborCommunicator::PrepareAndSendSyncLists                        0.000125      0.000175      0.000149  0.000036 
          geosx::NeighborCommunicator::UnpackGhosts                                   0.036664      0.072236      0.054206  0.013163 
            "Unpack FixedToManyElementRelation"                                       0.000848      0.001465      0.001157  0.000281 
            "Unpack OrderedVariableToManyElementRelation"                             0.001640      0.002810      0.002220  0.000539 
            geosx::ObjectManagerBase::UnpackGlobalMaps                                0.005305      0.007860      0.006587  0.001600 
          geosx::NeighborCommunicator::PrepareAndSendGhosts                           0.048518      0.095084      0.070621  0.017149 
        geosx::CommunicationTools::FindMatchedPartitionBoundaryObjects                0.000617      0.001152      0.000884  0.000215 
        geosx::CommunicationTools::AssignGlobalIndices                                0.015213      0.015721      0.015476  0.003758 
          geosx::EdgeManager::ExtractMapFromObjectForAssignGlobalIndexNumbers         0.000156      0.000295      0.000226  0.000055 
          geosx::ObjectManagerBase::ConstructGlobalToLocalMap                         0.000750      0.001393      0.001050  0.000255 
          geosx::FaceManager::ExtractMapFromObjectForAssignGlobalIndexNumbers         0.000183      0.000241      0.000213  0.000052 
      geosx::CellElementRegion::GenerateAggregates                                    0.000002      0.000003      0.000002  0.000001 
      geosx::DomainPartition::GenerateSets                                            0.000649      0.001160      0.000919  0.000223 
      geosx::NodeManager::SetEdgeMaps                                                 0.002346      0.002903      0.002629  0.000638 
      geosx::EdgeManager::BuildEdges                                                  0.002700      0.003651      0.003179  0.000772 
        geosx::populateMaps                                                           0.001213      0.001346      0.001289  0.000313 
        geosx::resizeEdgeToFaceMap                                                    0.000262      0.000423      0.000347  0.000084 
        geosx::createEdgesByLowestNode                                                0.005271      0.007171      0.006217  0.001510 
      geosx::NodeManager::SetFaceMaps                                                 0.002721      0.004421      0.003569  0.000867 
      geosx::FaceManager::BuildFaces                                                  0.003410      0.005515      0.004419  0.001073 
        geosx::populateMaps                                                           0.002560      0.004577      0.003595  0.000873 
        geosx::resizeFaceToNodeMap                                                    0.001775      0.002959      0.002395  0.000582 
        geosx::createFacesByLowestNode                                                0.005962      0.008473      0.007217  0.001753 
      geosx::NodeManager::SetElementMaps                                              0.000687      0.001169      0.000924  0.000224 
      geosx::ObjectManagerBase::ConstructGlobalToLocalMap                             0.000354      0.000605      0.000480  0.000117 
      geosx::InternalMeshGenerator::GenerateMesh                                      0.003162      0.005545      0.004314  0.001048
```

The total time spent in ComputeFaceNodalArea is **0.29%** of the runtime. @rrsettgast, your implementation is inside AssembleForceResidualDerivativeWrtTraction (0.06%), but the function does also something else.